modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,173 @@
+          tipParentsSize = tipParents.size
+          tipParentsToRemove = tipParents.filter {
+            case (_, block) =>
+              allTips.size >= tipsConfig.minimumTipsCount + tipParentsSize &&
+                (
+                  allTips.size > tipsConfig.maximumTipsCount + tipParentsSize ||
+                    block.usages.size >= tipsConfig.maximumTipUsages - 1
+                )
+          }
+          tipParentsToRemain = tipParents -- tipParentsToRemove.keySet
+          remainingTips = allTips -- tipParentsToRemove.keySet
+          minTipHeight = remainingTips.values.map(_.block.signed.value.height.coerce).minOption.getOrElse(0L)
+          addAsTipCondition = remainingTips.size < tipsConfig.maximumTipsCount && minTipHeight < hashedBlock.signed.value.height.coerce
+          _ <- updateNoTipParents(notTipParents, hashedBlock)
+          _ <- updateTipParentsToRemove(tipParentsToRemove.keySet.toList, hashedBlock)
+          _ <- updateTipParentsToRemain(tipParentsToRemain.keySet.toList, hashedBlock)
+          _ <- tryAddAsATip(addAsTipCondition, hashedBlock)
+        } yield ()
+
+      def pullTips(tipsCount: PosInt): F[Option[PulledTips]] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (_, AcceptedBlock(block, _, _, true)) => block }.toList
+            .sortBy(_.signed.value.height.coerce)
+          pulledTips = allTips match {
+            case tips if tips.size >= tipsCount =>
+              NonEmptyList
+                .fromList(tips.take(tipsCount))
+                .map(parents => block.PulledTips(parents.map(p => BlockReference(p.proofsHash, p.signed.value.height))))
+            case _ => None
+          }
+        } yield pulledTips
+
+      private def isBlockAccepted(blockReference: BlockReference): F[Boolean] =
+        blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[AcceptedBlock]))
+
+      private def updateNoTipParents(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, NotTipAcceptedBlockNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def updateTipParentsToRemove(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
+                (accepted.addUsage(hashedBlock).unsetAsTip.some, ().asRight)
+              case unexpected =>
+                (unexpected, TipBlockToUnsetAsTipNotFound(hash, unexpected).asLeft)
```

Comment:
same here

Reply:
I think this and the one below are good enough. The one above may be misleading so I will try to figure out the better name. This says that: a block that is a tip and was gonna be removed from tips wasn't found (when it should be found). And I think the name reflects that.

Reply:
We use `ToRemove` in 116 but `ToUnset` here. So either remove or unset. I would avoid mixing both. Name is fine for me as it fits `unsetAsTip`